File: python/Q210_course_schedule_ii.py
#
# @lc app=leetcode id=210 lang=python3
#
# [210] Course Schedule II
#


# @lc code=start
class Solution:
    def findOrder(self, numCourses: int, prerequisites: list[list[int]]) -> list[int]:
        indegrees: list[int] = [0] * numCourses
        adjacent_list: dict[int, list[int]] = {i: [] for i in range(numCourses)}

        for i, j in prerequisites:
            adjacent_list[j].append(i)
            indegrees[i] += 1

        leaves: list[int] = [i for i, v in enumerate(indegrees) if v == 0]

        finished: list[int] = []
        while leaves:
            node: int = leaves.pop()
            finished.append(node)

            for n in adjacent_list[node]:
                indegrees[n] -= 1
                if indegrees[n] == 0:
                    leaves.append(n)

        return finished if len(finished) == numCourses else []


# @lc code=end

File: rust/src/leetcode/Q210_course_schedule_ii.rs
/*
 * @lc app=leetcode id=210 lang=rust
 *
 * [210] Course Schedule II
 */

// @lc code=start
impl Solution {
    pub fn find_order(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> Vec<i32> {
        let mut indegrees: Vec<usize> = vec![0; num_courses as usize];
        let mut prerequisites_map: Vec<Vec<usize>> = vec![vec![]; num_courses as usize];

        for prerequisite in prerequisites {
            indegrees[prerequisite[0] as usize] += 1;
            prerequisites_map[prerequisite[1] as usize].push(prerequisite[0] as usize);
        }

        let mut current: Vec<usize> = indegrees
            .iter()
            .enumerate()
            .filter_map(|(course, &indegree)| (indegree == 0).then_some(course))
            .collect();

        let mut finished: i32 = 0;
        let mut result: Vec<i32> = Vec::new();
        while !current.is_empty() {
            finished += 1;

            let course: usize = current.pop().unwrap();
            result.push(course as i32);

            for &advanced_course in &prerequisites_map[course] {
                indegrees[advanced_course] -= 1;
                if indegrees[advanced_course] == 0 {
                    current.push(advanced_course);
                }
            }
        }

        if finished == num_courses {
            result
        } else {
            Vec::new()
        }
    }
}
// @lc code=end
